core/planner/planner.go

```diff
@@ -203,29 +155,55 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
 	// don't precondition longer than charging duration
 	precondition = min(precondition, requiredDuration)
 
+	// filter rates to planning window early for performance
+	rates = clampRates(rates, now, targetTime)
+
 	// reduce target time by precondition duration
 	targetTime = targetTime.Add(-precondition)
-	requiredDuration = max(requiredDuration-precondition, 0)
 
 	// separate precond rates, to be appended to plan afterwards
 	var precond api.Rates
 	if precondition > 0 {
 		rates, precond = splitPreconditionSlots(rates, targetTime)
+
+		// reduce required duration by precondition, skip planning if required
+		requiredDuration = max(requiredDuration-precondition, 0)
+		if requiredDuration == 0 {
+			return precond
+		}
 	}
 
 	// create plan unless only precond slots remaining
 	var plan api.Rates
-	if requiredDuration > 0 {
-		// sort rates by price and time
-		slices.SortStableFunc(rates, sortByCost)
+	if continuous {
+		// check if available tariff slots span is sufficient for sliding window algorithm
+		// verify that actual tariff data covers enough duration (may have gaps or start late)
+		if len(rates) > 0 {
+			start := rates[0].Start
+			if start.Before(now) {
+				start = now
+			}
+			end := rates[len(rates)-1].End
+			if end.After(targetTime) {
+				end = targetTime
+			}
 
+			// available window too small for sliding window - charge continuously from now to target
+			if end.Sub(start) < requiredDuration {
+				return continuousPlan(append(rates, precond...), now, targetTime.Add(precondition))
```

Comment:
Isn't that just

    return continuousPlan(rates, latestStart, targetTime)

? Wondering if that should have been covered above. How do we end up here?